apollo-router-core/src/cache.rs

```diff
@@ -76,50 +77,62 @@ where
         // miss a broadcast.
         drop(locked_cache);
 
-        match locked_wait_map.get_mut(&key) {
-            Some(waiter) => {
-                // Register interest in key
-                let mut receiver = waiter.subscribe();
-                drop(locked_wait_map);
-                // Our use case is very specific, so we are sure
-                // that we won't get any errors here.
-                let (recv_key, recv_value) = receiver.recv().await.expect(
-                    "the sender won't ever be dropped before all the receivers finish; qed",
-                );
-                debug_assert_eq!(recv_key, key);
-                recv_value
-            }
-            None => {
-                let (tx, _rx) = broadcast::channel(1);
-                locked_wait_map.insert(key.clone(), tx.clone());
-                drop(locked_wait_map);
-                // This is the potentially high duration operation where we ask our resolver to
-                // resolve the key (retrieve a value) for us
-                // No cache locks are held here
-                let value = self.resolver.retrieve(key.clone()).await;
-
-                // this is a separate block used to release the locks after editing the cache and wait map,
-                // but before broadcasting the value
-                {
-                    // Update our cache
-                    let mut locked_cache = self.cached.lock().await;
-                    locked_cache.put(key.clone(), value.clone());
-                    // Update our wait list
-                    let mut locked_wait_map = self.wait_map.lock().await;
-                    locked_wait_map.remove(&key);
+        loop {
+            match locked_wait_map.get_mut(&key) {
+                Some(weak_waiter) => {
+                    // Try to upgrade our weak Arc. If we can't, the sender must have
+                    // been cancelled, so remove the entry from the map and try again.
+                    let waiter = match Weak::upgrade(weak_waiter) {
+                        Some(waiter) => waiter,
+                        None => {
+                            locked_wait_map.remove(&key);
+                            continue;
+                        }
+                    };
+                    // Register interest in key
+                    let mut receiver = waiter.subscribe();
+                    drop(locked_wait_map);
+                    // Our use case is very specific, so we are sure
+                    // that we won't get any errors here.
+                    let (recv_key, recv_value) = receiver
+                        .recv()
+                        .await
+                        .expect("we ensured our receiver was valid before receiving; qed");
```

Comment:
we checked that the receiver was valid before subscribing, but if the first request is cancelled after that, and a new req comes in and removes the sender, `recv()` will return a `Err(RecvError::Closed)`: https://docs.rs/tokio/latest/tokio/sync/broadcast/error/enum.RecvError.html
so this `expect()` could panic